In [1]:
%%html
<style>
.h1_cell, .just_text {
    box-sizing: border-box;
    padding-top:5px;
    padding-bottom:5px;
    font-family: "Times New Roman", Georgia, Serif;
    font-size: 125%;
    line-height: 22px; /* 5px +12px + 5px */
    text-indent: 25px;
    background-color: #fbfbea; //#e6e6e6;
    
    padding: 10px;
}

hr { 
    display: block;
    margin-top: 0.5em;
    margin-bottom: 0.5em;
    margin-left: auto;
    margin-right: auto;
    border-style: inset;
    border-width: 2px;
}
</style>

<h1>
<center>
Module 1: Sign-up to be a wrangler!
</center>
</h1>
<div class=h1_cell>
In this module we will use a handful of features from the Python *`pandas`* library. I'll go through some basic clean-up steps using a spreadsheet/table taken from the sinking of the Titanic on its maiden voyage. The vernacular for the clean-up we will be doing is *data wrangling*. Common wisdom is that data wrangling can take up 70% of your entire effort. Life is messy. Data comes to us in both incomplete and erroneous forms. We have to deal with it.

</div>

<hr>
<h1>
Read in spreadsheet
</h1>
<div class=h1_cell>
<p>
For the first part of the course, we will be working on a problem called predictive analytics. The data we will be using to make predictions will be in spreadsheet form (I'll also call this *table* form).
<p>
We could read in the data to our own custom Python data-structure. And we might try that a little later. For now, we will use a very useful library called pandas. The pandas library was built with data wrangling in mind. It gives us a natural way to represent our data in table form.
<p>
I am going to use something called comma-separated values or csv as my raw file format. I like csv because you can use it to pass data around easily from things like Excel and google Sheets. And pandas knows how to read raw csv format and produce its own version called a dataframe.
<p>
I have the titanic data stored on google sheets. I used sheets to give me a url to the csv version of the file. Once I have that url, I can hand it to pandas and suck it in. Pretty dang cool.
<p>
BTW: it is convention to alias pandas as `pd`. It is also convention to use `df` to name a dataframe. I am using `titanic_table` in place of `df`.

</div>

In [2]:
import pandas as pd
url = 'https://docs.google.com/spreadsheets/d/1z1ycUZjJpmMWB4gXbhwRQ9B_qa42CwzAQkf82mLibxI/pub?output=csv'
titanic_table = pd.read_csv(url)

In [3]:
#I am setting the option to see all the columns of our table as we build it, i.e., it has no max.
pd.set_option('display.max_columns', None)

<hr>
<h1>
Explore
</h1>
<p>
<div class=h1_cell>

<ul>
<li>Use *`head`* to get general layout.</li>
<p>
<li>Find which columns have *`NaN`*s (empties) and how many.</li>
<p>
<li>Use *`describe`* method to see if any odd looking columns, e.g., more than 2 unqiue values for a binary column.</li>
</ul>
</div>

In [4]:
titanic_table.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

<div class=just_text>
<p>
There are a mixture of column types. Some have discrete values (e.g., `Pclass`, `Sex`, `Embarked`), some have continuous values (e.g., `Age`, `Far`e), and some are inbetween (e.g., `SibSp`, `Parch`). The `Name` column has unique values. The `Ticket` and `Cabin` columns are a bit of a hodge podge and will take further wrangling to make them useful.
<p>
Let's next see how many empties there are in each column.
<div>

In [7]:
titanic_table.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

<div class=just_text>
<p>
<ul>
<li>The `Age` column is a bit worrisome. It looks like a column that can be useful in prediction but has 177 empty values.
<p>
<li>The `Cabin` column has a lot of empties. I am dubious that the column as a whole will be useful. However, it might make sense to use the empty/non-empty question. For instance, maybe passengers with non-empty cabins were more likely to survive.
<p>
<li>The `Embarked` column has only 2 empties and that seems like something we can fill in.
<p>
</ul>
We will return to wrangling these empties shortly. For now, let's complete our initial exploration by using the `describe` method. I am using the value `all` to get discrete columns along with numeric (continuous) columns.
<div>

In [8]:
titanic_table.describe(include='all')

PassengerId    Survived      Pclass                       Name   Sex  \
count    891.000000  891.000000  891.000000                        891   891   
unique          NaN         NaN         NaN                        891     2   
top             NaN         NaN         NaN  Arnold-Franchi, Mr. Josef  male   
freq            NaN         NaN         NaN                          1   577   
mean     446.000000    0.383838    2.308642                        NaN   NaN   
std      257.353842    0.486592    0.836071                        NaN   NaN   
min        1.000000    0.000000    1.000000                        NaN   NaN   
25%      223.500000    0.000000    2.000000                        NaN   NaN   
50%      446.000000    0.000000    3.000000                        NaN   NaN   
75%      668.500000    1.000000    3.000000                        NaN   NaN   
max      891.000000    1.000000    3.000000                        NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare        Cabin  \
count   714.000000  891.000000  891.000000     891  891.000000          204   
unique         NaN         NaN         NaN     681         NaN          147   
top            NaN         NaN         NaN  347082         NaN  C23 C25 C27   
freq           NaN         NaN         NaN       7         NaN            4   
mean     29.699118    0.523008    0.381594     NaN   32.204208          NaN   
std      14.526497    1.102743    0.806057     NaN   49.693429          NaN   
min       0.420000    0.000000    0.000000     NaN    0.000000          NaN   
25%      20.125000    0.000000    0.000000     NaN    7.910400          NaN   
50%      28.000000    0.000000    0.000000     NaN   14.454200          NaN   
75%      38.000000    1.000000    0.000000     NaN   31.000000          NaN   
max      80.000000    8.000000    6.000000     NaN  512.329200          NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

<div class=just_text>
<p>
Not all of this information is useful. I believe we can use the mean of certain continuous columns a little later. For continuous columns, we might also make use of the standard-deviation (std). Also the min and max.
<p>
For the discrete columns, I can see the mode of the column using top. I am also interested in the number of unique values for discrete columns. For instance, I would expect the `Sex` column to have 2 unique values (`male`, `female`). And it has a value of 2. So all good. For the `Embarked` column, I know that there are 3 unique values. And I see a 3 above. So all good.
<p>
What if I saw a number greater than 2 for `Sex` or 3 for `Embarked`? It is an indicator that a value has been mistyped. For instance, maybe the value `Male` was entered instead of `male`. Case counts. If I want to see all the unqiue values in a column, I can do so.
</div>

In [9]:
titanic_table.Embarked.unique()  # or titanic_table['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

<hr>
<h1>Drop useless columns</h1>
<p>
<div class=h1_cell>
<p>

As we move through the wrangling process, it may be clear that certain columns are unneeded. Either they carry no useful information or they duplicate information we have in other columns. In the case of 
`PassengerId`, it is a type of index. But we already have an index so don't need another. Let's drop it.
<p>
Note I am using `axis=1` to say I am working on columns and not rows (`axis=0`). I want to drop an entire column.
<p>
Also note that I am doing `titanic_table = titanic_table...`. In general I am using pandas methods that return a copy of the table with the changes I made. I am reassigning `titanic_table` to the copy. I suppose I could keep a lot of variables around like `titanic_table_1`, `titanic_table_2`, etc. Never overwrite a variable. But I find trying to manage such a name space clumsy. It is true my way does not allow you to roll back to a prior version of the table. But you can "roll forward" by just executing all of the cells from the top of the ntoebook to get to a specific state.
</div>

In [10]:
titanic_table = titanic_table.drop(['PassengerId'], axis=1)

In [11]:
titanic_table.head()  # Should see drop of column

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S

<hr>
<h1>Outliers</h1>
<p>
<div class=h1_cell>
<p>

One simple approach to outliers is to use the 3-sigma rule: https://www.encyclopediaofmath.org/index.php/Three-sigma_rule. Anything below `(mean - 3sigma)` or above `(mean + 3sigma)` should be looked at carefully. Just for practice, let's check out the `Fare` column.
</div>

In [12]:
sigma3_fare =  titanic_table['Fare'].std() * 3 # calculate std * 3
mean_fare =   titanic_table['Fare'].mean()# calculate mean
low_fare = mean_fare - sigma3_fare
high_fare = mean_fare + sigma3_fare
print((low_fare, high_fare))

(-116.87607782296804, 181.2844937601173)


In [13]:
#Build sub-table of all rows with low_fare outliers - should be the empty table - no rows in negative range
titanic_table.loc[titanic_table['Fare'] < low_fare]

Empty DataFrame
Columns: [Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [14]:
#Build sub-table of all rows with high_fare outliers - should have 20 rows
titanic_table.loc[titanic_table['Fare'] > high_fare]

Survived  Pclass                                               Name  \
27          0       1                     Fortune, Mr. Charles Alexander   
88          1       1                         Fortune, Miss. Mabel Helen   
118         0       1                           Baxter, Mr. Quigg Edmond   
258         1       1                                   Ward, Miss. Anna   
299         1       1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
311         1       1                         Ryerson, Miss. Emily Borie   
341         1       1                     Fortune, Miss. Alice Elizabeth   
377         0       1                          Widener, Mr. Harry Elkins   
380         1       1                              Bidois, Miss. Rosalie   
438         0       1                                  Fortune, Mr. Mark   
527         0       1                                 Farthing, Mr. John   
557         0       1                                Robbins, Mr. Victor   
679         1       1                 Cardeza, Mr. Thomas Drake Martinez   
689         1       1                  Madill, Miss. Georgette Alexandra   
700         1       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
716         1       1                      Endres, Miss. Caroline Louise   
730         1       1                      Allen, Miss. Elisabeth Walton   
737         1       1                             Lesurer, Mr. Gustave J   
742         1       1              Ryerson, Miss. Susan Parker "Suzette"   
779         1       1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
27     male  19.0      3      2     19950  263.0000      C23 C25 C27        S  
88   female  23.0      3      2     19950  263.0000      C23 C25 C27        S  
118    male  24.0      0      1  PC 17558  247.5208          B58 B60        C  
258  female  35.0      0      0  PC 17755  512.3292              NaN        C  
299  female  50.0      0      1  PC 17558  247.5208          B58 B60        C  
311  female  18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
341  female  24.0      3      2     19950  263.0000      C23 C25 C27        S  
377    male  27.0      0      2    113503  211.5000              C82        C  
380  female  42.0      0      0  PC 17757  227.5250              NaN        C  
438    male  64.0      1      4     19950  263.0000      C23 C25 C27        S  
527    male   NaN      0      0  PC 17483  221.7792              C95        S  
557    male   NaN      0      0  PC 17757  227.5250              NaN        C  
679    male  36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
689  female  15.0      0      1     24160  211.3375               B5        S  
700  female  18.0      1      0  PC 17757  227.5250          C62 C64        C  
716  female  38.0      0      0  PC 17757  227.5250              C45        C  
730  female  29.0      0      0     24160  211.3375               B5        S  
737    male  35.0      0      0  PC 17755  512.3292             B101        C  
742  female  21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
779  female  43.0      0      1     24160  211.3375               B3        S

<div class=just_text>
<p>Several fares look suspicious to me, e.g., rows 258, 679 and 737. The mean of this column is 32. Did someone really pay over 500? It is interesting to note that all passengers are in `Pclass` 1, i.e., first-class. Maybe they gouged first class passengers.
<p>
I won't do any further wrangling with outliers. Some options you might consider for dealing with an outlier in future:
<ul>
<li>Change to NaN. Treat like a missing value. Then will fall under however you handle missing values.</li>
<p>
<li>Change to mean.</li>
<p>
<li>Change to (mean + 3sigma) - e, i.e., epsilon below the bar.</li>
</div>

<hr>
<h1>Handle empties</h1>
<p>
<div class=h1_cell>
<p>
I am going to show you how to deal with empty values in two ways. As a reminder, an empty value is represented by a `NaN` in a cell. Note that a `NaN` value is not the string NaN. It is a special symbol that pandas knows about.

<p><b>Approach 1</b>. For each existing column of interest, add a new column that tracks empties. If the existing column has an empty then set the new column to 1. If the existing column has a non-empty value, set the new column to 0.
<p><b>Approach 2</b>. For an existing continuous column, create a new column that is a copy except that empties have been filled in with a legitimate value (e.g., mean).
<p>
Notice in both cases, I am not modifying the existing columns. I like to keep them just in case I need to fall back on them. In essence, I'm adding new information to the table, not changing what already exists.
<p>
Why follow approach 1, i.e., why keep empties? Because they may have some inferential value. Perhaps a missing Age signifies that the passenger did not survive. If they did survive, their Age information could be filled in later. This is a shaky conjecture but the general idea is good - don't throw away information!
<p>
Of course, you can take both approaches: add a column that tracks empties and add a column that fills them in.
</div>


<h2>Let's start with a continuous column</h2>
<p>
<div class=h1_cell>
Let's start with a continuous column. We will target the `Age` column. First we will build a new column that shows empties and then build a new column that fills empties with mean.
<p>
The operation of this code is to supply the functional argument with one row at a time: `titanic_table.apply(`. You can treat the row as a dictionary and pull out values from it using column names.
<p>
Note that I cannot use `if row['Age'] != 'NaN'`. `NaN` is a special symbol and you need to check for it with `pd.isnull`.
</div>

In [15]:
titanic_table['no_age'] = titanic_table.apply(lambda row: 1 if pd.isnull(row.Age) else 0, axis=1)
titanic_table.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

   no_age  
0       0  
1       0  
2       0  
3       0  
4       0

<div class=just_text>
Create a new column with empties filled with mean - call it `filled_age`.
</div>

In [16]:
# Get the mean of Age - I also could look at describe info above to get it
mean_age = titanic_table['Age'].mean()

In [17]:
# Build the new column
titanic_table['filled_age'] = titanic_table.apply(lambda row: mean_age if pd.isnull(row.Age) else row.Age, axis=1)
titanic_table.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

   no_age  filled_age  
0       0        22.0  
1       0        38.0  
2       0        26.0  
3       0        35.0  
4       0        35.0

<h2>Let's now look at a categorical column</h2>
<p>
<div class=h1_cell>

We will focus on `Embarked`. To get a new column that shows empties, I am going to jump ahead a bit and use *one-hot encoding* (ohe). One byproduct of ohe is a column that shows empties like we want.
<p>
The general idea is to create a new column for each distinct value in `Embarked`. There are 3 legitimate values and also the empty value. So we have 4 values to encode. Sounds complicated but pandas makes it easy. I'll demonstrate below.
<p>
Note that `dummy_na=True` says I want a column that shows empties. I am also prefixing each new column name with `emb`. If I did not use this prefix, I would get new columns C, Q, S. The problem is if other columns use any of these values. If they did, I could run into naming conflicts. It is easier to just consistenly add a prefix that makes column names unique.
</div>

In [18]:
one_hot_emb = pd.get_dummies(titanic_table['Embarked'],prefix='emb',dummy_na=True)
titanic_table = titanic_table.join(one_hot_emb)
titanic_table.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

   no_age  filled_age  emb_C  emb_Q  emb_S  emb_nan  
0       0        22.0      0      0      1        0  
1       0        38.0      1      0      0        0  
2       0        26.0      0      0      1        0  
3       0        35.0      0      0      1        0  
4       0        35.0      0      0      1        0

<div class=just_text>

If we decide later that we want to fill-in the 2 empty values in `Embarked`, a good choice is to use the mode. There is a mode function as part of pandas. But I find it complicated to use. I could just look at the top value in the `describe` output to see the mode. Instead, I'll ask for the count of all values in the column and just eyeball it.
</div>

In [19]:
titanic_table['Embarked'].value_counts()  # easier to decipher than mode()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

<div class=just_text>
The `S` value is the big winner. If I decide to fill empties, I'll use S as the filler. 
</div>

<hr>
<h1>Bin continuous columns</h1>
<p>
<div class=h1_cell>

The general idea is to break a continuous column, e.g., `Age`, into discrete bins. We have the choice of defining the bins ourselves or letting pandas do it for us. To do it ourselves, we would build a list of boundaries. Then use that list with the pd.cut method. I like this example from StackOverflow for describing how flags can be used to set different bin boundaries.
</div>

<div class=just_text>
<p>
You read `(x,y]` as value is greater-than x and less-than-or-equal to y.
<p>
You read `[x,y)` as value is greater-than-or-equal to x and less-than y. You use the cut parameters to get what you want.
<p>
You can also use a number in place of the list. Let's see what happens if I allow cut to decide how to choose the 2 bins.
</div>

In [20]:
test = pd.DataFrame({'days': [0,20,30,31,45,60]})

test['range1'] = pd.cut(test.days, 2, include_lowest=True)  # I am using 2 where normally the boundary list would be
#30 value is in [30, 60) group
test['range2'] = pd.cut(test.days, 2, right=False)
#30 value is in (0, 30] group
test['range3'] = pd.cut(test.days, 2)
print (test)

   days          range1         range2         range3
0     0  (-0.061, 30.0]    [0.0, 30.0)  (-0.06, 30.0]
1    20  (-0.061, 30.0]    [0.0, 30.0)  (-0.06, 30.0]
2    30  (-0.061, 30.0]  [30.0, 60.06)  (-0.06, 30.0]
3    31    (30.0, 60.0]  [30.0, 60.06)   (30.0, 60.0]
4    45    (30.0, 60.0]  [30.0, 60.06)   (30.0, 60.0]
5    60    (30.0, 60.0]  [30.0, 60.06)   (30.0, 60.0]


<div class=just_text>
Notice that `cut` sets the bin boundaries such that all values are included, i.e., no NaNs produced. Let's go ahead and divide the `filled_age` (`Age` with empties filled with mean) column into 3 bins. And let's provide a label for each bin.
</div>

In [21]:
bins = 3  # will divide into 3 evenly spaced bins
bin_names = ['Child', 'Adult', 'Senior']
titanic_table['age_bin'] = pd.cut(titanic_table['filled_age'], bins, labels=bin_names)

In [22]:
titanic_table['age_bin'].value_counts()

Adult     522
Child     319
Senior     50
Name: age_bin, dtype: int64

<div class=just_text>
The `pd.cut` method only focuses on even splits of the range of the column. You can see the 3 bins have different numbers of rows. There is another method, `pd.qcut`, that focuses on evenly populated bins. It chooses bin boundaries so that the rows are evenly distributed among bins. Let's see what it looks like.

In [23]:
test = pd.DataFrame({'days': [0,20,30,31,32,33,45,60]})  # I added a couple more values.

test['range1'] = pd.qcut(test.days, 2)
print (test)

   days          range1
0     0  (-0.001, 31.5]
1    20  (-0.001, 31.5]
2    30  (-0.001, 31.5]
3    31  (-0.001, 31.5]
4    32    (31.5, 60.0]
5    33    (31.5, 60.0]
6    45    (31.5, 60.0]
7    60    (31.5, 60.0]


<h2>Ready for ohe</h2>
<p>
<div class=h1_cell>

We now have `age_bin` as a discrete valued column. We can use one-hot encoding on it, right?
</div>

In [24]:
one_hot_age = pd.get_dummies(titanic_table['age_bin'],prefix='age',dummy_na=False)  # false because should not have empties
titanic_table = titanic_table.join(one_hot_age)
titanic_table.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

   no_age  filled_age  emb_C  emb_Q  emb_S  emb_nan age_bin  age_Child  \
0       0        22.0      0      0      1        0   Child          1   
1       0        38.0      1      0      0        0   Adult          0   
2       0        26.0      0      0      1        0   Child          1   
3       0        35.0      0      0      1        0   Adult          0   
4       0        35.0      0      0      1        0   Adult          0   

   age_Adult  age_Senior  
0          0           0  
1          1           0  
2          0           0  
3          1           0  
4          1           0

<h2>One downside of OHE</h2>
<p>
<div class=h1_cell>
<p>
There can be issues with OHE when using something other than decision tree models; it can make you do more work later. One fix is to use one of the values as a "reference value" and leave it out. The idea is that if you have k values and only produce k-1 columns, you can use a 0 in all those k-1 columns to infer that the reference value is true. For instance, if we made C the reference value, then when S = 0 and Q = 0 and empty = 0 then it must be the case that C = 1.
<p>
And in fact, `get_dummies` gives you an easy way to generate k-1 new columns using the `drop_first=True` parameter (its default is False). But I am choosing not to use it. Why? Because I believe it makes it harder for us humans to understand what is going on. And has no effect on decision trees. And even if we were using a regression model instead, you can still find ways around it. See these posts if interested: https://stats.stackexchange.com/q/231285, https://stats.stackexchange.com/a/224055. And a longer article: https://arxiv.org/pdf/1511.05728.pdf.
</div>

<hr>
<h1>Write dataframe to file</h1>
<p>
<div class=h1_cell>
<p>
You could recreate the table to this point by rerunning all the cells above. Alternatively, you can check-point the table you have so that you can simply load it next time and start from there. I am going to do the latter.
<p>
I make use of dropbox as my cloud storage area. I like it because it integrates with the native file system. Allows me to access my data on different computers. I'll take a week 1 check-point here. You can choose your own `file_path` value.
</div>

In [25]:
import os

week = 1

home_path =  os.path.expanduser('~')

file_path = '/Dropbox/cis399_ds1_f17/notebook_history/' #on my to-do list to rename folder w19

file_name = 'titanic_wrangled_w'+str(week)+'.csv'

titanic_table.to_csv(home_path + file_path + file_name, index=False)


<h1>Read it back in</h1>
<p>
<div class=h1_cell>
<p>

Read it back in just to make sure of round-trip
</div>

In [26]:
test_table = pd.read_csv(home_path + file_path + file_name)
test_table.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

   no_age  filled_age  emb_C  emb_Q  emb_S  emb_nan age_bin  age_Child  \
0       0        22.0      0      0      1        0   Child          1   
1       0        38.0      1      0      0        0   Adult          0   
2       0        26.0      0      0      1        0   Child          1   
3       0        35.0      0      0      1        0   Adult          0   
4       0        35.0      0      0      1        0   Adult          0   

   age_Adult  age_Senior  
0          0           0  
1          1           0  
2          0           0  
3          1           0  
4          1           0

<hr>
<h1>Write code to script</h1>
<p>
<div class=h1_cell>
<p>

Finally, you will want to follow these same steps for other data we work with, e.g., test data, actual data from the user. The code below uses the exclamation-point to run a shell command from the notebook. The result is that you get a python file that is runnable as a script. You will probably have to clean up the file but it is a good first start. We will be using this a little later in the course.
</div>

In [27]:
# Save the entire notebook as a python file (i.e., a script)
!jupyter nbconvert --to script module1_pandas.ipynb

[NbConvertApp] Converting notebook module1_pandas.ipynb to script
[NbConvertApp] Writing 18471 bytes to module1_pandas.py


<hr>
<h1>Next up</h1>
<div class=h1_cell>

We will next use our wrangled table to build some simple decision trees.
</div>